cmake/object_library.cmake

```diff
@@ -0,0 +1,218 @@
+#
+# cmake/environment.cmake
+#
+# The MIT License
+#
+# Copyright (c) 2022 TileDB, Inc.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+# -------------------------------------------------------
+#
+# Environment `object_library`
+#
+# -------------------------------------------------------
+#
+# Properties:
+# - target: The name of a CMake object library target
+# - sources: A list of sources for the target.
+#   - Must not be empty
+# - object_library_dependencies: Other object libraries the target depends upon
+# - link_library_dependencies: Ordinary libraries the target depends upon
+#
+
+# -------------------------------------------------------
+# Preliminaries
+# -------------------------------------------------------
+include_guard()
+include(environment)
+
+#
+# Module-scope log level
+#
+set(TileDB_Environment_object_library_Log_Level VERBOSE)
+message(${TileDB_Environment_object_library_Log_Level} ">>> object_library.cmake begin >>>")
+
+#
+# The environment object_library
+#
+commence(define_environment "object_library")
+    set_this_property(commence_hook TileDB_Environment_object_library_begin)
+    set_this_property(conclude_hook TileDB_Environment_object_library_end)
+conclude(define_environment)
```

Comment:
I have been using two macros in experimental for the specific cases of header-only object libraries -- one macro for the setup to build the object library and the `compile_<library>_main` and one macro for the setup to build a unit test `unit_<library>`.  A comment and a question.
- I think the `object_library` environment should "just work" for a header-only object library.
- Would it be useful to have a `unit_test` environment?

Reply:
* CMake does not support header-only object libraries. On the other hand, ordinary libraries can carry header dependencies, so it's not much of a restriction in practice. It does, unfortunately, mean that there's a small exposure about whether a library carries sources or not.
  * CMake object libraries are specifically written to be containers of object code. Header-only libraries don't generate such object code, so that's fundamental. The way CMake works, object libraries do need to specify their list of objects at some point, so the distinction matters to CMake. Whether this is the best idea is another question, but also one out of scope. We have limited capacity to rewrite CMake fundamentals.
  * It might be possible for "our" object libraries to cheat this with something like a header-only declaration, which would then (behind the scenes) add a dependency on an "empty" object library, which means an object library that compiles an empty compilation unit but still emits (empty) object code. That would satisfy the link requirements of a CMake object library. On the other hand, I'm not at all sure that's a good idea. Pushing the boundaries of arbitrary compiler tool chains with stuff like this seems imprudent. My sense is that it's better to leave header-only libraries out of the object library system.

* A unit test environment is next. I had originally planned to do it in this PR, but after working through object libraries, I realized a combined PR would be too much. There's more variation in the unit tests as well.